### PR TITLE
BFD-1570: Write test to verify access log format

### PR DESCRIPTION
**Pull request by JeremyIR**
**JIRA Ticket:**
[BFD-1570](https://jira.cms.gov/browse/BFD-1570)

**User Story or Bug Summary:**

User Story:

As a BFD engineer, I want logs in the correct format, so that they can be used for alerting.

---

### What Does This PR Do?

Write unit and integration tests to check the format of access logs 
### What Should Reviewers Watch For?

Do the tests and regex look thorough and simple enough to prevent any issues with our access.log format?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

 N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java [hunk 1]
```diff
@@ -29,6 +33,32 @@
 public final class DataServerLauncherAppIT {
   /** The POSIX signal number for the <code>SIGTERM</code> signal. */
   private static final int SIGTERM = 15;
+  /** Regex for access log entries */
+  private static final String accessLogPattern =
+      "^(\\S+) \\S+ \\\"([^\\\"]*)\\\" \\[([^\\]]+)\\] \\\"([A-Z]+) ([^ \\\"]+) HTTP\\/[0-9.]+\\\" \\\"([^ \\\"]+)\\\" ([0-9]{3}) ([0-9]+|-) ([0-9]+|-) (\\S+) ([0-9]+|-) \\[([^\\]]+)\\] ([0-9]+|-) \\\"([^\\\"]*)\\\" ([0-9]+|-) \\\"([^\\\"]*)\\\" ([0-9]+|-) \\\"([^\\\"]*)\\\" (\\S+)";
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java:38]: wondering if you can break this up a little bit so that the pieces are easier to think about individually. perhaps like this:

```java
private static final String accessLogPattern =
   "^(\\S+) " + \\ Description of first field
   "\\S+ " + \\ Description of second field
   ...
```

Stringbuilder would be an option as well using append() for each piece.

If you come up with something nice you could also use it below for the positive and negative tests.

Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java:38]: wondering if you can break this up a little bit so that the pieces are easier to think about individually. perhaps like this:

```java
private static final String accessLogPattern =
   "^(\\S+) " + \\ Description of first field
   "\\S+ " + \\ Description of second field
   ...
```

Stringbuilder would be an option as well using append() for each piece.

If you come up with something nice you could also use it below for the positive and negative tests.

File: apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java [hunk 2]
```diff
@@ -29,6 +33,32 @@
...
+    String goodLine1 =
+        "127.0.0.1 - \"CN=client-local-dev\" [07/Mar/2022:18:43:15 +0000] \"GET / HTTP/1.1\" \"?null\" 200 26 22000 - - [-] - \"-\" - \"-\" - \"-\" -";
+    String goodline2 =
+        "10.252.14.216 - \"EMAILADDRESS=gomer.pyle@adhocteam.us, CN=BlueButton Root CA, OU=BlueButton on FHIR API Root CA, O=Centers for Medicare and Medicaid Services, L=Baltimore, ST=Maryland, C=US\" [01/Oct/2021:23:10:01 -0400] \"GET /v1/fhir/Coverage/?startIndex=0&_count=10&_format=application%2Fjson%2Bfhir&beneficiary=Patient%2F587940319 HTTP/1.1\" \"?startIndex=0&_count=10&_format=application%2Fjson%2Bfhir&beneficiary=Patient%2F587940319\" 200 2103 23 3b3e2b30-232f-11ec-9b9f-0a006c0cb407 1 [2021-10-02 03:10:01.104125] 11770 \"-\" 32 \"Evidation on behalf of Heartline\" 79696 \"-\" patientId:587940319";
+    String badLine =
+        "127.0.0.1 - \"CN=client-local-dev\" [[07/Mar/2022:18:43:15 +0000]] \"GET / HTTP/1.1\" \"?null\" 200 26 22000 - - [-] - \"-\" - \"-\" - \"-\" -";
+    String badLine2 =
+        "127.0.0.1 - \"CN=client-local-dev\" [07/Mar/2022:18:43:15 +0000] \"GET / HTTP/1.1\" \"?null\" 2004 26 22000 - - [-] - \"-\" - \"-\" - \"-\" -";
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java:53]: can you write a comment about what aspect of these lines makes them bad

Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java:53]: can you write a comment about what aspect of these lines makes them bad

File: apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java [hunk 3]
```diff
@@ -29,6 +33,32 @@
...
+  /**
+   * Verifies the regex for valdiating our access log entries adequately avoids edge cases that
+   * could break our alerts, which depend on logs
+   */
+  @Test
+  public void checkAccessLogFormat() {
+    String goodLine1 =
+        "127.0.0.1 - \"CN=client-local-dev\" [07/Mar/2022:18:43:15 +0000] \"GET / HTTP/1.1\" \"?null\" 200 26 22000 - - [-] - \"-\" - \"-\" - \"-\" -";
+    String goodline2 =
+        "10.252.14.216 - \"EMAILADDRESS=gomer.pyle@adhocteam.us, CN=BlueButton Root CA, OU=BlueButton on FHIR API Root CA, O=Centers for Medicare and Medicaid Services, L=Baltimore, ST=Maryland, C=US\" [01/Oct/2021:23:10:01 -0400] \"GET /v1/fhir/Coverage/?startIndex=0&_count=10&_format=application%2Fjson%2Bfhir&beneficiary=Patient%2F587940319 HTTP/1.1\" \"?startIndex=0&_count=10&_format=application%2Fjson%2Bfhir&beneficiary=Patient%2F587940319\" 200 2103 23 3b3e2b30-232f-11ec-9b9f-0a006c0cb407 1 [2021-10-02 03:10:01.104125] 11770 \"-\" 32 \"Evidation on behalf of Heartline\" 79696 \"-\" patientId:587940319";
```
Comment (christrotteradhoc) [inline apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java:49]: looks like they were a couple of lines pulled from a real access.log which is nice -- can you document that in a comment?

**Review by christrotteradhoc — COMMENTED** [review]
Nice work. I have a few suggestions that could improve readability

**Review by christrotteradhoc — APPROVED** [review]

File: apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java [hunk 4]
```diff
@@ -29,6 +34,102 @@
...
+          .add("\"([^\"]*)\"") // BlueButton-User
+          .add("(\\S+)") // BlueButton-BeneficiaryId
+          .toString();
+
+  /**
+   * Verifies the regex for valdiating our access log entries adequately avoids edge cases that
+   * could break our alerts, which depend on logs
+   */
+  @Test
+  public void checkAccessLogFormat() {
```
[the remark this answers is not in the document]
Comment (christrotteradhoc) [reply apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java:66]: We discussed this some and the other side of this is that this regex only exists for the IT test that uses it to verify output (and that must be an IT test because it has to start the server process and generate a log file). So putting that regex in another class and then putting the unit tests for the regex in yet another class is a substantial increase in scope and indirection. Trading that complexity for having a unit test in the IT class seemed reasonable to me.

[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-server/bfd-server-launcher/src/test/java/gov/cms/bfd/server/launcher/DataServerLauncherAppIT.java:66]: Hm I see the place that uses this regex to parse the logs is external to BFD, so this is kind of a weird case in terms of testing. Worried we need to remember to update this when/if we change the pattern for parsing the logs, since the place we use pattern this won't break this test if we change it. Not sure how to connect the two, but this at least offers some coverage, so I'm cool with it.

**Review by lsmitchell — APPROVED** [review]
